File: app/src/main/java/com/example/asiantech/playmusic/fragment/ArtistFragment.java
package com.example.asiantech.playmusic.fragment;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;

import com.example.asiantech.playmusic.MainActivity_;
import com.example.asiantech.playmusic.OnItemListener;
import com.example.asiantech.playmusic.R;
import com.example.asiantech.playmusic.adapter.ArtistAdapter;
import com.example.asiantech.playmusic.model.Artist;
import com.example.asiantech.playmusic.model.Song;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.ViewById;

import java.util.ArrayList;

@EFragment(R.layout.artist_fragment)
public class ArtistFragment extends Fragment implements OnItemListener{
    private ArrayList<Song> mListSong;
    private ArrayList<Artist> mListArtist;
    @ViewById(R.id.recycleViewListArtist)
    RecyclerView mRecyclerListArtist;

    @AfterViews
    void afterView() {
        Log.d("TAG ARTIST_FRAGMENT", "afterView");
        if (getActivity() instanceof MainActivity_) {
            mListSong = ((MainActivity_) getActivity()).getMListSong();
        }
        mListArtist = new ArrayList<>();
        for (int i = 0; i < mListSong.size(); i++) {
            boolean exists = false;
            for (int j = 0; j < mListArtist.size(); j++) {
                if (mListSong.get(i).getArtistId() == mListArtist.get(j).getArtistId()) {
                    exists = true;
                    break;
                }
            }
            if (!exists) {
                mListArtist.add(new Artist(mListSong.get(i).getArtistId(),
                        mListSong.get(i).getArtist()));
            }
        }
        ArtistAdapter adapter = new ArtistAdapter(getContext(), mListArtist, this);
        mRecyclerListArtist.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecyclerListArtist.setAdapter(adapter);
    }

    @Override
    public void onItemClick(int position) {
        ArtistDetailFragment_ fragment = new ArtistDetailFragment_();
        Bundle bundle = new Bundle();
        bundle.putLong("artistId", mListArtist.get(position).getArtistId());
        fragment.setArguments(bundle);
        FragmentTransaction transaction = getActivity().getSupportFragmentManager()
                .beginTransaction();
        transaction.add(R.id.rlContainer, fragment);
        transaction.addToBackStack(null);
        transaction.commit();
    }
}

File: app/src/main/java/com/example/asiantech/playmusic/fragment/ArtistFragment.kt
package com.example.asiantech.playmusic.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.asiantech.playmusic.MainActivity
import com.example.asiantech.playmusic.OnItemListener
import com.example.asiantech.playmusic.R
import com.example.asiantech.playmusic.adapter.ArtistAdapter
import com.example.asiantech.playmusic.model.Artist
import com.example.asiantech.playmusic.model.Song
import kotlinx.android.synthetic.main.artist_fragment.*

/**
 * ArtistFragment.
 *
 * @author HoaHT
 */
class ArtistFragment : Fragment(), OnItemListener {

    private var listSong = ArrayList<Song>()
    private var listArtist = ArrayList<Artist>()
    private var adapter: ArtistAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (activity is MainActivity) {
            listSong = (activity as MainActivity).getListSong()
        }
        for (i in listSong.indices) {
            val exists = listArtist.indices.any {
                listSong[i].artistId == listArtist[it].artistId
            }
            if (!exists) {
                val song = listSong[i]
                listArtist.add(Artist(song.artistId, song.artist))
            }
        }
        adapter = ArtistAdapter(context, listArtist, this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return LayoutInflater.from(context).inflate(R.layout.artist_fragment, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recycleViewListArtist.layoutManager = LinearLayoutManager(context)
        recycleViewListArtist.adapter = adapter
    }

    override fun onItemClick(position: Int) {
        val transition = activity.supportFragmentManager.beginTransaction()
        val fragment = ArtistDetailFragment()
        val bundle = Bundle()
        bundle.putLong(BaseDetailFragment.KEY_ALBUM_ID, listArtist[position].artistId)
        fragment.arguments = bundle
        transition.add(R.id.rlContainer, fragment).addToBackStack(null).commit()
    }

}